### ISSUE TITLE
String data, right truncation

### ISSUE THREAD

**Issue by srish13** [mkleehammer/pyodbc#423]
### Environment
- OS: ubuntu:16.04
- Python: 2.7
- pyodbc:  4.0.23 
- DB:  SQL Server 12
- unixodbc-dev=2.3
-  Microsoft ODBC Driver 13 (13.1.9.2-1)

### Issue
On inserting a string containing emoji characters into an NVARCHAR(MAX) column in MSSQL, I get the following error:
```(pyodbc.DataError) ('22001', u'[22001] [Microsoft][ODBC Driver 13 for SQL Server]String data, right truncation ```.
#### However, it works perfectly fine on windows and mac machine.

On changing python's default encoding to UTF-8, like
```
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
```
I don't get above error. But inserted data are  different symbols.
``` 
sql = "insert into test(user,  category)  values ('abcd', '🌜🌜🌜')" 
cur.execute(sql)
```
ð??¤ð??¤ð??¤ - Inserted in the database. 

```
sql = "insert into test(user, category) values (?, ?)"
cur.execute(sql, ['abcd', '🌜🌜🌜'])
```
If I give parameterized query like above,  ' ???????' are inserted.

- I even tried to change the encoding of pyodbc to (refering this https://github.com/mkleehammer/pyodbc/wiki/Unicode):
1 - con.setencoding(unicode, encoding='utf-8')
2- con.setdecoding(pyodbc.SQL_CHAR, encoding='latin1', to=str)
    con.setencoding(unicode, encoding='raw')
3- cnxn.setdecoding(pyodbc.SQL_CHAR, encoding='latin1', to=str)
cnxn.setencoding(str, encoding='latin1')
 In all the avove cases '??????' is inserted instead of emoticon. 

#### I even tried Microsoft ODBC Driver 17 for SQL Server. But nothing changed.

**Comment by v-chojas** [comment]
What is the exact version of unixODBC? Run `odbcinst -j` to get that information.

Post an [ODBC trace](https://www.simba.com/blog/odbc-troubleshooting-tracing/ ) to help with debugging the problem further.

**Comment by v-chojas** [comment]
Enable tracing in odbcinst.ini and subsequent connections will be traced.

We cannot determine what is happening without an ODBC trace.

**Comment by srish13** [comment]
odbcinst -j output
```
unixODBC 2.3.1
DRIVERS............: /etc/odbcinst.ini
SYSTEM DATA SOURCES: /etc/odbc.ini
FILE DATA SOURCES..: /etc/ODBCDataSources
USER DATA SOURCES..: /home/srpatel/.odbc.ini
SQLULEN Size.......: 8
SQLLEN Size........: 8
SQLSETPOSIROW Size.: 8
```
###ODBC Trace 
```
[ODBC][6069][1531742452.316771][__handles.c][460]
                Exit:[SQL_SUCCESS]
                        Environment = 0x155d2d0
[ODBC][6069][1531742452.316799][SQLSetEnvAttr.c][189]
                Entry:
                        Environment = 0x155d2d0
                        Attribute = SQL_ATTR_ODBC_VERSION
                        Value = 0x3
                        StrLen = 4
[ODBC][6069][1531742452.316807][SQLSetEnvAttr.c][363]
                Exit:[SQL_SUCCESS]
[ODBC][6069][1531742452.316813][SQLAllocHandle.c][375]
                Entry:
                        Handle Type = 2
                        Input Handle = 0x155d2d0
[ODBC][6069][1531742452.316819][SQLAllocHandle.c][493]
                Exit:[SQL_SUCCESS]
                        Output Handle = 0x1545fc0
[ODBC][6069][1531742452.316956][SQLDriverConnectW.c][290]
                Entry:
                        Connection = 0x1545fc0
                        Window Hdl = (nil)
                        Str In = [DRIVER={ODBC Driver 13 for SQL Server};PORT=1433;SERVER=miq-dev.database.windows.net;DATABASE=xxxx;UID=xxxxx;PWD=.xxx][length = 134]
                        Str Out = (nil)
                        Str Out Max = 0
                        Str Out Ptr = (nil)
                        Completion = 0
                UNICODE Using encoding ASCII 'ISO8859-1' and UNICODE 'UCS-2LE'

[ODBC][6069][1531742452.393563][SQLDriverConnectW.c][910]
                Exit:[SQL_SUCCESS_WITH_INFO]
                        Connection Out [[NULL]]
"odbc.log" 299L, 9609C
```

**Comment by v-chojas** [comment]
Need the whole ODBC trace file (attach it), not just first page copy-pasted from vi.

**Comment by srish13** [comment]
I have got an interesting behavior. My application is using Flask and SQLAlchemy. When I insert emoticon data using sqlalchemy & pyodbc = 4.0.22. Its gets inserted without any extra change. 
How with 4.0.17( my current application uses this version) and 4.0.23, I get **String data, right truncation** error. 
However direct queries like these, insert junk symbols irrespect of 4.0.22 or 4.0.23.
```
sql = "insert into test(user,  category)  values ('abcd', '🌜🌜🌜')" 
cur.execute(sql)
```
How its working with sqlalchemy + 4.0.22 and not with 4.0.23.

**Comment by v-chojas** [comment]
> Need the whole ODBC trace file (attach it), not just first page copy-pasted from vi.
Do this for both cases so we can compare.

**Comment by srish13** [comment]
ODBC traces for below code, using pyodbc= 4.0.22 and 4.0.23
[odbc_4.0.22.log](https://github.com/mkleehammer/pyodbc/files/2202605/odbc_4.0.22.log)
[odbc_4.0.23.log](https://github.com/mkleehammer/pyodbc/files/2202606/odbc_4.0.23.log)
`# -*- coding: utf-8 -*-
import pyodbc
import sys
reload(sys)
sys.setdefaultencoding('utf-8')


if __name__ == '__main__':
    sql_connect_str = "DRIVER={ODBC Driver 13 for SQL Server};PORT=1433;SERVER=xxxx.database.windows.net;DATABASE=xxx;UID=xxxx;PWD=xxxx"
    print str('s1') + " connection string = " + str(sql_connect_str)
    con = pyodbc.connect(sql_connect_str, autocommit=True)
    cur = con.cursor()
    a = "🌜🌜"
    sql = "insert into user_move_detail(user_parse_id, to_shard, from_shard, state, category)" \
          " values ('testing_4.0.23', 's0', 's1', 'VALID_USER', '🌜🌜')"
    cur.execute(sql)
    print 'insert done'
    con.commit()
    con.close()
`

**Comment by srish13** [comment]
Traces using sqlalchemy to insert emoticon data with pyodbc version = 4.0.22 and pyodbc = 4.0.23.
SQLALCHEMY + pyodbc=4.0.22 = Emoticon are inserted
SQLALCHEMY + pyodbc=4.0.23 = DataError: (pyodbc.DataError) ('22001', u'[22001] [Microsoft][ODBC Driver 13 for SQL Server]String data, right truncation

[odbc_4.0.22.log](https://github.com/mkleehammer/pyodbc/files/2205685/odbc_4.0.22.log)
[odbc_4.0.23.log](https://github.com/mkleehammer/pyodbc/files/2205687/odbc_4.0.23.log)

**Comment by srish13** [comment]
Uploading again- "Traces using sqlalchemy to insert emoticon data".  The code is inserting data in a table named 'vehicle' using sqlalchemy.
[odbc_4.0.22.log](https://github.com/mkleehammer/pyodbc/files/2208837/odbc_4.0.22.log)
[odbc_4.0.23.log](https://github.com/mkleehammer/pyodbc/files/2208839/odbc_4.0.23.log)

**Comment by srish13** [comment]
What is the update on this issue? The above traces are most recent traces with pyodbc version = 4.0.22 and 4.0.23.

**Comment by srish13** [comment]
@v-chojas Using pyodbc=4.0.22. Solves the problem of emoticon data. However, pyodbc = 4.0.22 introduced another typecasting issue -  Invalid character value for cast specification (0) (SQLExecDirectW)') . From some other parts of my application. 
Going through StackOverflow and Google, I realized that 4.0.22 is not stable and have few issues.  
Wondering why emoticon is supported in 4.0.22 and not in 4.0.23???

**Comment by v-makouz** [comment]
The truncation error happens because when binding variables pyODBC sets the column size to number of characters (so for u'🌜🌜' that would be 2), but the driver expects the number of UTF-16 code units, and the moons are higher Unicode characters so each one is 2 code units, thus the column size should be set to 4.

It works in 4.0.22 because that version uses automatic detection of SQL types and column sizes using SQLDescribeParam to bind variables

I made a pull request with one way of solving this problem:
https://github.com/mkleehammer/pyodbc/pull/434

P.S.
The reason direct query like:
sql = "insert into test(user,  category)  values ('abcd', '🌜🌜🌜')"
inserts junk is because the Unicode string should have an N in the front, like:
sql = "insert into test(user,  category)  values ('abcd', N'🌜🌜🌜')"

More info here:
https://docs.microsoft.com/en-us/sql/t-sql/data-types/nchar-and-nvarchar-transact-sql?view=sql-server-2017

Relevant quote:
"Prefix Unicode character string constants with the letter N. Without the N prefix, the string is converted to the default code page of the database. This default code page may not recognize certain characters."

**Comment by srish13** [comment]
@v-makouzThe fix in the PR solves the problem. Thanks :)

**Comment by mkleehammer** [comment]
@v-chojas Is this something you can confirm with someone on the SQL Server driver team?

I think the explanation of the difference is correct, but I'm not sure which piece of code has a bug.

The MS docs on [ColumnSize](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/column-size?view=sql-server-2017) are pretty clear that ColumnSize should be the number of characters, which I would have expected to be unrelated to encoding.

A one-character string with a [half-moon](http://www.fileformat.info/info/unicode/char/1f31c/index.htm) is 4 bytes long when encoded in UTF-8 or in UTF-16.  However, the number of *code points* is certainly still one.  I would have assumed that "characters" meant code points.

Are we sure this isn't a bug in the SQL Server driver?

If I understand PR #434 correctly, it would set the column size to 2 for the example one-character string.  I get it is because each UCS2 "character" is 2-bytes, but that is not true of UTF-16 which is a variable length encoding.

This really sounds like SQL Server is being a bit lazy and treating UTF-16 like UCS2.  When I ran an insert in SQL Server Management Studio, `select len(v)` returned 2, which really isn't correct.  Though, I was unable to get it to display as a half-moon, so I'm not entirely sure that was a good test.). Is the issue perhaps SQL Server does not support Unicode characters that are larger than 2-bytes?

Right now, databases set to use UTF-8 are not failing, but I think they could with this change.  (Though most seem much more tolerant of input, so we might get lucky.). If it really is just SQL Server when using wchar_t, perhaps we just need a special check for utf-16.

Also, I'd recommend we look at using the encoding constants.  There are some constants for the different UTF-16 which is safer than looking at the string.  After all, the encoding is [officially "utf_16"](https://docs.python.org/2/library/codecs.html) in Python with aliases like "U16" and "utf16" that might get missed by looking for strings.

(Beware of using ctype.  It can be set to 2 when writing UTF-8 to a wchar_t buffer.  That's just what some of the databases require.)

**Comment by mkleehammer** [comment]
My tests with PostgreSQL and MySQL also worked, so I'm ready to move forward.  (I'd be more comfortable if the ODBC specification, not just the SQL Server ODBC docs, specifically referred to wchar_t counts when using UTF-16, but I think you are right.)

@v-makouz How about the 2nd part.  Can we change it to look at the `enc.optenc` flag instead of `enc.name`?  That way all of the code that looks for the different combinations that mean "utf-16" are in one place (`SetTextEncCommon`).

**Comment by v-chojas** [comment]
@mkleehammer that page is part of the ODBC spec, MS just reorganised the docs recently. If you'd like to see the same paragraph in an earlier version of the spec, see the second half of page 183 in https://cdn.simba.com/wp-content/uploads/2016/03/ODBC_specification.pdf

**Comment by mkleehammer** [comment]
@v-chojas Perfect, thanks!

**Comment by v-makouz** [comment]
@mkleehammer I've modified the PR to use enc.optenc, I ran the tests for SQL Server, MySQL, PostgreSQL and Oracle and didn't see any regressions, I also made sure the moons are still inserted correctly after that change

**Comment by mkleehammer** [comment]
Released in 4.0.24.  Thanks everyone.
